### PR TITLE
JSAB2-96 Connect resource front end to back end

### PR DESCRIPTION
**Pull request by GuJialu**

### CODE REVIEW

**Review by HONGTAO-T — APPROVED** [review]

**Review by HanqingZ — APPROVED** [review]

File: frontend/src/Resources/index.jsx [hunk 1]
```diff
@@ -84,19 +87,48 @@ ResourceView.propTypes = {
...
+  const updateResourcesInterval = setInterval(() => dispatch(updateResources()), 1000);
+
+  useEffect(() => {
+    dispatch(fetchResources());
+    return () => clearInterval(updateResourcesInterval);
+  }, []);
+
+  if (error) {
+    return (
+      <Text>{`Oops, ${error.message}`}</Text>
```
Comment (evasimonyi) [inline frontend/src/Resources/index.jsx:107]: @GuJialu you said you'll handle this when connection the frontend to the backend
from the action creator you are sending back the status code as the payload when there's an error, so the user will see the status code here? please check

**Review by evasimonyi — CHANGES_REQUESTED** [review]
Right now the calculations and the resource updating is in the reducer, so there is no communication with the backend, only for the first time when you are fetching the mocked resources —> this way the frontend is not in sync with the backend. On the backend we wouldn’t store the actual amount. Right now everything is happening on the frontend, but the task says exactly the opposite: to connect the frontend to the backend.

Please refactor your code, and put the whole logic and calculations to the backend, and just communicate with the backend, dispatching an action from the frontend in every minute is perfect. 

I know you tried this solution for optimizing the performance, but sometimes it can not be more important than maintainability and testability. Also right now our goal is to practice as much as you can, and the HTTP communication is a crucial part in every application.

File: frontend/src/Resources/index.jsx (hunk 1 above)
Comment (GuJialu) [reply frontend/src/Resources/index.jsx:107]: I think the error will eventually be send to the error page as a prop, and then the error page will decided what to show to the user depend on the error. Currently it is just showing the error handling is working.

**Review by HONGTAO-T — APPROVED** [review]

**Review by HanqingZ — APPROVED** [review]

**Review by HanqingZ — APPROVED** [review]

**Review by kyya — APPROVED** [review]

File: backend/routers/resources.js [hunk 2]
```diff
@@ -7,20 +7,33 @@ const mockedResources = {
...
 
+function getResourceAmount(initialAmount, generation, updatedAt) {
+  const date = new Date();
+  const minutesPassed = (date.getTime() - updatedAt) / 60000;
+  return Math.round(initialAmount + generation * minutesPassed);
+}
+
 function getResources(req, res) {
-  return res.status(200).send(mockedResources);
+  const caulatedResources = {
```
Comment (evasimonyi) [inline backend/routers/resources.js]:
```suggestion
  const calculatedResources = {
```

File: backend/routers/resources.js [hunk 3]
```diff
@@ -7,20 +7,33 @@ const mockedResources = {
...
+    resources: mockedResources.resources.map((resource) => ({
+      type: resource.type,
+      amount: getResourceAmount(resource.amount, resource.generation, resource.updatedAt),
+      generation: resource.generation,
+    })),
+  };
+  return res.status(200).send(caulatedResources);
```
Comment (evasimonyi) [inline backend/routers/resources.js]:
```suggestion
  return res.status(200).send(calculatedResources);
```

**Review by evasimonyi — CHANGES_REQUESTED** [review]
you had 2 typos, if you fix them, I'll merge, no need for reviews again, if you don't modify anything else:)
also you can accept these here

**Review by evasimonyi — APPROVED** [review]

**Pull request merged**